### PR TITLE
Propagate group information when analyzing row filter and column masks

### PR DESCRIPTION
**Pull request by Praveen2112**
## Description

Fixes https://github.com/trinodb/trino/issues/15583

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix query failure when column masking or row filter involves columns. ({issue}`15583`)
```

### CODE REVIEW

**Review by ksobolew — COMMENTED** [review]
I think this will also break when the user has roles enabled. But this is a hairy problem. I believe the SPI should allow the access control to return a full `Identity` in `ViewExpression` like we do for views with `SECURITY DEFINER`. Right now, for row filters and column masks, if there are roles involved, we have to guess which ones are enabled for the user. (Any user, not just the current session.)

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 1]
```diff
@@ -954,6 +958,34 @@ public void testPropertiesAccessControlIsSkippedWhenUsingDefaults()
         assertAccessAllowed("CREATE MATERIALIZED VIEW mock.default.new_materialized_view AS SELECT 1 a");
     }
 
+    @Test
+    public void testAccessControlWithGroupsAndColumnMask()
+    {
+        String securityConfigFile = getResource("access_control_rules_with_groups.json").getPath();
+        systemAccessControl.set(new FileBasedSystemAccessControl.Factory()
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:965]: Can we use something like `DenySetPropertiesSystemAccessControl` to mimic the behavior `FileBasedSystemAccessControl`? Then the existing tests can be moved to `BaseFileBasedConnectorAccessControlTest`. WDYT?

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 2]
```diff
@@ -954,6 +958,34 @@ public void testPropertiesAccessControlIsSkippedWhenUsingDefaults()
...
+    }
+
+    @Test
+    public void testAccessControlWithGroupsAndRowFilter()
+    {
+        String securityConfigFile = getResource("access_control_rules_with_groups.json").getPath();
+        systemAccessControl.set(new FileBasedSystemAccessControl.Factory()
+                .create(ImmutableMap.of(SECURITY_CONFIG_FILE, securityConfigFile)));
+
+        assertAccessDenied("SELECT * FROM nation", "Cannot select from table .*");
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:982]: can we have full error message?

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 3]
```diff
@@ -954,6 +958,34 @@ public void testPropertiesAccessControlIsSkippedWhenUsingDefaults()
...
     }
 
+    @Test
+    public void testAccessControlWithGroupsAndColumnMask()
+    {
+        String securityConfigFile = getResource("access_control_rules_with_groups.json").getPath();
+        systemAccessControl.set(new FileBasedSystemAccessControl.Factory()
+                .create(ImmutableMap.of(SECURITY_CONFIG_FILE, securityConfigFile)));
+
+        assertAccessDenied("SELECT comment FROM orders", "Cannot select from table .*");
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:968]: can we have full error message?

**Review by kokosing — APPROVED** [review]
Suggestions are optional.

**Comment by kokosing** [comment]
Thank you for fixing this.

**Comment by kokosing** [comment]
> I believe the SPI should allow the access control to return a full Identity in ViewExpression like we do for views with SECURITY DEFINER. 

This would be a big change. Notice that `io.trino.spi.connector.ConnectorViewDefinition#getOwner` return just an user name like it is in `ViewExpression`. Only `io.trino.metadata.SystemSecurityMetadata#getViewRunAsIdentity` returns `Identity` at the moment and this is not part of SPI.

We can consider using `Identity` in SPI here, but:
 - I don't think it should be the way to fix the bug. As bugfixing is tactical and that change would be strategic. So we should not rush and change SPI being under time pressure.
 - We should change everything view related (masks, filters and views) in one go.

**Comment by kokosing** [comment]
> OK, so there should be an analogous method for filters and masks. Or maybe we can reuse the one for views? As you say, filters and masks are view-like, after all.

TBH I don't know yet. Storing entire `Identity` in the plugin does not sound like a good idea too. Notice that `Identity` has plenty of strange things. Groups that come from authentication, roles used in some catalogs, system roles, extra credentials. Maybe we should use `TrinoPrincipal`? But then we don't have groups.

Maybe the way it is implemented today (including this PR) is good enough? Notice that we get user and all its groups so we can populate needed roles if needed.

Anyway, I wanted to say it is complex. And it is not 100% sure if cure is better than sickness here.

**Comment by ksobolew** [comment]
> Maybe we should use `TrinoPrincipal`? But then we don't have groups.

Groups are part of the `Identity`, but they are not managed by access controls, so they have to be added later anyway. Also, doesn't this limit us to either the user name or a single role name?

**Comment by guyco33** [comment]
I approve that it fix the issue.
Thanks @Praveen2112 for this quick fix

**Comment by Praveen2112** [comment]
@kokosing AC

**Comment by Praveen2112** [comment]
>  Or maybe we can reuse the one for views? As you say, filters and masks are view-like, after all.

For row filters maybe we could reuse it but for column masks we pass additional information like column name and column type so it would be dependent on it.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 4]
```diff
@@ -110,6 +113,7 @@
 public class TestAccessControl
         extends AbstractTestQueryFramework
 {
+    private static final String GROUP = "mock";
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:116]: It does not have to be constant.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 5]
```diff
@@ -998,4 +1026,43 @@ private static void checkProperties(Map<?, ?> properties)
             }
         }
     }
+
+    private static final class AllowSpecificGroupSystemAccessControl
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:1030]: Hm... sorry. Once you wrote code I realized that you most likely use `TestingAccessControlManager`, see:
 - `TestingAccessControlManager#rowFilter`
 - `TestingAccessControlManager#columnMask`
 - `TestingAccessControlManager#denyIdentityTable`

**Comment by Praveen2112** [comment]
@kokosing AC

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 6]
```diff
@@ -954,6 +956,41 @@ public void testPropertiesAccessControlIsSkippedWhenUsingDefaults()
         assertAccessAllowed("CREATE MATERIALIZED VIEW mock.default.new_materialized_view AS SELECT 1 a");
     }
 
+    @Test
+    public void testAccessControlWithGroupsAndColumnMask()
+    {
+        getQueryRunner().getGroupProvider().setUserGroups(ImmutableMap.of(getSession().getUser(), ImmutableSet.of("group")));
+
+        TestingAccessControlManager accessControlManager = getQueryRunner().getAccessControl();
+
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:965]: nit: not needed empty lines, here, one above and one below

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 7]
```diff
@@ -954,6 +956,41 @@ public void testPropertiesAccessControlIsSkippedWhenUsingDefaults()
...
+        assertAccessAllowed("SELECT comment FROM orders");
+    }
+
+    @Test
+    public void testAccessControlWithGroupsAndRowFilter()
+    {
+        getQueryRunner().getGroupProvider().setUserGroups(ImmutableMap.of(getSession().getUser(), ImmutableSet.of("group")));
+
+        TestingAccessControlManager accessControlManager = getQueryRunner().getAccessControl();
+
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:983]: nit: not needed empty lines, here, one above and one below

**Review by kokosing — APPROVED** [review]

**Review by ksobolew — APPROVED** [review]
OK, let's fix this. @kokosing says that Trino is not so role-focused, so we'll see if anyone runs into this with roles.

**Pull request merged**